### PR TITLE
Features/proxy

### PR DESCRIPTION
**Pull request by kovpaka**

### CODE REVIEW

File: Proxy/Program.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        {
+            YesterdayRate yr = new YesterdayRate();
+            Console.WriteLine(yr.getRate().ToString());
+            int answer = 0;
+            CachedYesterdayRate cachedYesterdayRate = new CachedYesterdayRate(yr);
+            while (answer != 1)
+            {
+                Console.WriteLine(cachedYesterdayRate.getRate().ToString());
+                Console.WriteLine("Input 1 to exit:");
+                answer = int.Parse(Console.ReadLine());
```
Comment (siarheikarabko) [inline Proxy/Program.cs:17]: What if a letter is pressed?

File: .gitignore [hunk 2]
```diff
@@ -3,4 +3,5 @@
 _ReSharper.*
 bin
 obj
-packages
\ No newline at end of file
+packages
+/.vs/design-patterns/DesignTimeBuild
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply .gitignore:7]: Changed the .gitignore at the develop branch

File: Proxy/Program.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace Proxy
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            YesterdayRate yr = new YesterdayRate();
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/Program.cs]: I see

File: Proxy/Rate.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿
+namespace Proxy
+{
+    class Rate
+    {
+        public string Dollar { get; set; }
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/Rate.cs]: Yes, you're right. Fixed

File: Proxy/CachedYesterdayRate.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        private Rate cacheRate;
+        private IYesterdayRate yesterdayRate;
+
+        public CachedYesterdayRate(IYesterdayRate yesterdayRate)
+        {
+            this.yesterdayRate = yesterdayRate;
+            SetTimer();
+        }
+
+        public Rate getRate()
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs]: Fixed

File: Proxy/CachedYesterdayRate.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        private void ClearCache()
+        {
+            cacheRate = null;
+        }
+
+        private void SetTimer()
+        {
+            timer = new Timer(interval);
+            timer.Elapsed += OnTimedEvent;
+            timer.AutoReset = true;
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs:47]: Done

File: Proxy/CachedYesterdayRate.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Timers;
+
+namespace Proxy
+{
+    class CachedYesterdayRate : IYesterdayRate
+    {
+        private const int interval = 100000;
+        private Timer timer;
+        private Rate cacheRate;
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs]: Fixed

File: Proxy/Program.cs (hunk 1 above)
Comment (kovpaka) [reply Proxy/Program.cs]: That's reasonable. Was added check for input

File: Proxy/CachedYesterdayRate.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            timer = new Timer(interval);
+        }
+
+        public Rate GetRate()
+        {
+            if (cachedRate == null)
+            {
+                Rate rate = yesterdayRate.GetRate();
+                AddToCache(rate);
+                if(!timer.Enabled)
```
Comment (siarheikarabko) [inline Proxy/CachedYesterdayRate.cs:25]: I don't think it's possible that the timer has been already started when a new value is added to the cache.

File: Proxy/Program.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        static void Main()
+        {
+            var yRate = new YesterdayRate();
+            Console.WriteLine(yRate.GetRate().ToString());
+            var cachedYesterdayRate = new CachedYesterdayRate(yRate);
+            while(true)
+            {
+                Console.WriteLine("Input 1 to continue:");
+                bool check = int.TryParse(Console.ReadLine(), out int answer);
```
Comment (siarheikarabko) [inline Proxy/Program.cs:15]: You could initialize `answer` with `0` before passing it to the parse function and then just check the value whether it's equal to `1`

File: Proxy/YesterdayRate.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Proxy
+{
+    class YesterdayRate : IYesterdayRate
+    {
+        public Rate GetRate()
+        {
+            var yesterdayRate = new Rate()
+            {
+                Dollar = double.Parse(DateTime.Now.Hour.ToString() + DateTime.Now.Minute.ToString()),
```
Comment (siarheikarabko) [inline Proxy/YesterdayRate.cs:11]: You could use there the custom formatting 
https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings

File: Proxy/CachedYesterdayRate.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (siarheikarabko) [reply Proxy/CachedYesterdayRate.cs:47]: Please, review

File: Proxy/Program.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+
+namespace Proxy
+{
+    class Program
+    {
+        static void Main()
+        {
+            var yRate = new YesterdayRate();
```
Comment (siarheikarabko) [inline Proxy/Program.cs:9]: Review the variable name

File: Proxy/CachedYesterdayRate.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs]: Already done for sure

File: Proxy/CachedYesterdayRate.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        private void ClearCache()
+        {
+            cachedRate = null;
+        }
+
+        private void SetTimer()
+        {
+
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs]: fixed

File: Proxy/CachedYesterdayRate.cs (hunk 8 above)
Comment (kovpaka) [reply Proxy/CachedYesterdayRate.cs]: No more timer, it has been replaced by the time field

File: Proxy/Program.cs (hunk 9 above)
Comment (kovpaka) [reply Proxy/Program.cs]: Done

File: Proxy/YesterdayRate.cs (hunk 10 above)
Comment (kovpaka) [reply Proxy/YesterdayRate.cs]: Fixed

File: Proxy/Program.cs (hunk 11 above)
Comment (kovpaka) [reply Proxy/Program.cs]: Fixed

File: Proxy/Program.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+            Console.WriteLine(yesterdayRate.GetRate().ToString());
+
+            var cachedYesterdayRate = new CachedYesterdayRate(yesterdayRate);
+            int answer = 0;
+            while (answer != 1)
+            {
+                Console.WriteLine(cachedYesterdayRate.GetRate().ToString());
+
+                Console.WriteLine("Input 1 to exit:");
+                if (int.TryParse(Console.ReadLine(), out int userAnswer))
```
Comment (siarheikarabko) [inline Proxy/Program.cs:19]: You could pass `answer` directly.
```
int answer = 0;
while (answer != 1) 
{
    ...
    int.TryParse(Console.ReadLine(), out answer);
}
```

Comment (kovpaka) [reply Proxy/Program.cs]: I don't know about it. Fixed.

**Pull request merged**